File: packages/omnivad/src/streaming-vad.ts
/**
 * High-level streaming VAD: PCM chunks in, completed segments out.
 *
 * Wraps OmniStreamVAD (model inference) and OmniStreamSegmenter (state machine)
 * into a single object so callers don't have to glue them together manually.
 *
 * Typical usage:
 *
 *   const vad = await OmniStreamingVAD.create();
 *   for (const chunk of pcmChunks) {                // Int16Array, any length
 *     for (const { start, end } of vad.process(chunk)) {
 *       console.log(`speech: ${start.toFixed(2)}s -> ${end.toFixed(2)}s`);
 *     }
 *   }
 *   for (const { start, end } of vad.flush()) {
 *     console.log(`speech (tail): ${start.toFixed(2)}s -> ${end.toFixed(2)}s`);
 *   }
 *
 * If you also need per-frame confidence, use OmniStreamVAD + OmniStreamSegmenter
 * directly to share inference output across multiple consumers.
 */

import type { ModelSource } from "./types.js";
import { OmniStreamVAD } from "./stream-vad.js";
import {
  OmniStreamSegmenter,
  type StreamSegment,
  type StreamSegmenterConfig,
} from "./stream-segmenter.js";

/** Combined config: ModelSource + StreamSegmenterConfig (segmenter applies
 *  the same threshold to its smoothed-prob comparison). */
export interface StreamingVADConfig extends ModelSource, StreamSegmenterConfig {
  /** Override the threshold passed to the underlying stream VAD. Defaults
   *  to whatever the segmenter uses (default 0.4) so the VAD's is_speech
   *  bit is consistent with the segmenter's view. */
  vadThreshold?: number;
}

export class OmniStreamingVAD {
  /** omni_stream_vad_process emits at most one frame per call regardless of
   *  input size, so we split incoming chunks into 160-sample sub-frames
   *  internally. (Codex flagged this gotcha during the design review.) */
  private static readonly STREAM_VAD_HOP = 160; // 10ms @ 16kHz

  private readonly vad: OmniStreamVAD;
  private readonly segmenter: OmniStreamSegmenter;
  private totalSamplesSeen = 0;
  private closed = false;

  private constructor(vad: OmniStreamVAD, segmenter: OmniStreamSegmenter) {
    this.vad = vad;
    this.segmenter = segmenter;
  }

  static async create(options: StreamingVADConfig = {}): Promise<OmniStreamingVAD> {
    const segmenterThreshold = options.threshold ?? 0.4;
    const vadThreshold       = options.vadThreshold ?? segmenterThreshold;

    const vad = await OmniStreamVAD.create({
      modelUrl:        options.modelUrl,
      modelData:       options.modelData,
      speechThreshold: vadThreshold,
    });
    const segmenter = await OmniStreamSegmenter.create({
      threshold:        segmenterThreshold,
      smoothWindowSize: options.smoothWindowSize,
      minSpeechFrames:  options.minSpeechFrames,
      minSilenceFrames: options.minSilenceFrames,
      maxSpeechFrames:  options.maxSpeechFrames,
    });
    return new OmniStreamingVAD(vad, segmenter);
  }

  /** Push one PCM chunk of arbitrary length. Returns 0+ completed segments. */
  process(pcmChunk: Int16Array): StreamSegment[] {
    if (this.closed) throw new Error("OmniStreamingVAD has been closed.");
    this.totalSamplesSeen += pcmChunk.length;

    const out: StreamSegment[] = [];
    const hop = OmniStreamingVAD.STREAM_VAD_HOP;
    for (let offset = 0; offset < pcmChunk.length; offset += hop) {
      const sub = pcmChunk.subarray(offset, offset + hop);
      const result = this.vad.processFrame(sub);
      if (result) {
        for (const seg of this.segmenter.processFrame(result.confidence)) out.push(seg);
      }
    }
    return out;
  }

  /** Emit any in-progress segment at end-of-stream. */
  flush(): StreamSegment[] {
    if (this.closed) throw new Error("OmniStreamingVAD has been closed.");
    return this.segmenter.flush(this.totalSamplesSeen);
  }

  // ---- State queries -----------------------------------------------------

  get isInSpeech(): boolean {
    return !this.closed && this.segmenter.isInSpeech;
  }

  get activeStart(): number | null {
    return this.closed ? null : this.segmenter.activeStart;
  }

  get totalSamples(): number {
    return this.totalSamplesSeen;
  }

  // ---- Lifecycle ---------------------------------------------------------

  reset(): void {
    if (this.closed) return;
    this.segmenter.reset();
    // OmniStreamVAD does not expose reset on the JS side; recreating is the
    // only safe path for a full reset of model state. For the typical
    // long-audio use-case, reset on the segmenter alone is sufficient.
    this.totalSamplesSeen = 0;
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.segmenter.close();
    // OmniStreamVAD doesn't expose explicit close in the public API; it's
    // owned by the GC. We hold no other resources to release here.
  }
}

File: omnivad/streaming_vad.py
"""High-level streaming VAD: audio chunks in, completed segments out.

Wraps :class:`OmniStreamVAD` (model inference) and :class:`OmniStreamSegmenter`
(segment state machine) into a single object so callers don't have to glue
them together manually.

Typical usage::

    from omnivad import OmniStreamingVAD
    import numpy as np

    vad = OmniStreamingVAD()                 # default config
    pcm = np.fromfile("speech.pcm", dtype=np.int16)
    chunk_size = 160                          # 10ms @ 16kHz

    for i in range(0, len(pcm), chunk_size):
        chunk = pcm[i : i + chunk_size]
        for start, end in vad.process(chunk):
            print(f"speech: {start:.2f}s -> {end:.2f}s")

    # End-of-stream: flush trailing in-progress segment (if any)
    for start, end in vad.flush():
        print(f"speech (tail): {start:.2f}s -> {end:.2f}s")

If you also need per-frame confidence, use OmniStreamVAD + OmniStreamSegmenter
directly — that's the supported way to share the inference output between
multiple consumers without doing it twice.
"""

from __future__ import annotations

from typing import List, Optional, Tuple

import numpy as np

from omnivad.stream_segmenter import OmniStreamSegmenter
from omnivad.stream_vad import OmniStreamVAD


class OmniStreamingVAD:
    """Audio-in → segments-out helper that owns a stream VAD + segmenter pair."""

    def __init__(
        self,
        model_path: Optional[str] = None,
        *,
        threshold: float = 0.5,
        smooth_window_size: int = 5,
        min_speech_secs: float = 0.20,
        min_silence_secs: float = 0.20,
        max_chunk_secs: float = 30.0,
    ):
        # Note: stream VAD uses its own threshold for is_speech but the
        # segmenter re-decides based on raw confidence + smoothing, so the
        # VAD threshold here mainly affects the StreamResult.is_speech bit
        # we don't use in this convenience wrapper.
        self._vad = OmniStreamVAD(model_path, threshold=threshold)
        self._segmenter = OmniStreamSegmenter(
            threshold=threshold,
            smooth_window_size=smooth_window_size,
            min_speech_secs=min_speech_secs,
            min_silence_secs=min_silence_secs,
            max_chunk_secs=max_chunk_secs,
        )
        self._total_samples_seen: int = 0

    # --------------------------------------------------------------------- #
    #  Streaming                                                             #
    # --------------------------------------------------------------------- #

    # Stream VAD's hop length: each omni_stream_vad_process() call advances
    # by exactly one frame. Feeding larger chunks in one call would silently
    # drop intermediate frames, so we split internally.
    _STREAM_VAD_HOP = 160  # 10ms @ 16kHz

    def process(self, pcm_chunk: np.ndarray) -> List[Tuple[float, float]]:
        """Feed one PCM chunk of arbitrary length (int16 or float32).
        Returns 0+ completed segments emitted during this call."""
        chunk = np.asarray(pcm_chunk)
        self._total_samples_seen += len(chunk)

        out: List[Tuple[float, float]] = []
        hop = self._STREAM_VAD_HOP
        for offset in range(0, len(chunk), hop):
            sub = chunk[offset : offset + hop]
            result = self._vad.process(sub)
            if result is not None:
                out.extend(self._segmenter.process_frame(result.confidence))
        return out

    def flush(self) -> List[Tuple[float, float]]:
        """Emit any in-progress segment at end-of-stream."""
        return self._segmenter.flush(self._total_samples_seen)

    # --------------------------------------------------------------------- #
    #  State queries                                                         #
    # --------------------------------------------------------------------- #

    @property
    def is_in_speech(self) -> bool:
        return self._segmenter.is_in_speech

    @property
    def active_start(self) -> Optional[float]:
        return self._segmenter.active_start

    @property
    def total_samples_seen(self) -> int:
        return self._total_samples_seen

    # --------------------------------------------------------------------- #
    #  Lifecycle                                                             #
    # --------------------------------------------------------------------- #

    def reset(self) -> None:
        """Reset both the VAD model state and the segmenter state machine."""
        self._vad.reset()
        self._segmenter.reset()
        self._total_samples_seen = 0

    def close(self) -> None:
        if self._segmenter:
            self._segmenter.close()
        if self._vad:
            self._vad.close()

    def __del__(self):
        try:
            self.close()
        except Exception:
            pass

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()
